### PR TITLE
Fix compilation of Jacobian with for loops

### PR DESCRIPTION
**Pull request by rmoyard**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new functions and code must be clearly commented and documented.

- [ ] Ensure that code is properly formatted by running `make format`.
      The latest version of black and `clang-format-14` are used in CI/CD to check formatting.

- [ ] All new features must include a unit test.
      Integration and frontend tests should be added to [`frontend/test`](../frontend/test/),
      Quantum dialect and MLIR tests should be added to [`mlir/test`](../mlir/test/), and
      Runtime tests should be added to [`runtime/tests`](../runtime/tests/).

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: frontend/catalyst/jit.py [hunk 1]
```diff
@@ -384,6 +384,7 @@ def compute_jvp(self, primals, tangents):
         # Optimization: Do not compute Jacobians for arguments which do not participate in
         #               differentiation.
         argnums = []
+        tangents, _ = tree_flatten(tangents)
```
Comment (rmoyard) [inline frontend/catalyst/jit.py:387]: Added for the jax version in the test to run

Comment (dime10) [reply frontend/catalyst/jit.py:387]: Oh, didn't we want to swap this implementation in Python for the `JVP` op? 😅

Comment (rmoyard) [reply frontend/catalyst/jit.py:387]: Yes I think we wanted but I don't recall why we stopped, but there was a reason I think.

File: mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp [hunk 2]
```diff
@@ -183,7 +183,7 @@ FailureOr<func::FuncOp> HybridGradientLowering::cloneCallee(PatternRewriter &rew
                 funcOp.walk([&](func::CallOp callOp) {
                     if (callOp.getCallee() == qnode.getName()) {
                         PatternRewriter::InsertionGuard insertionGuard(rewriter);
-                        rewriter.setInsertionPointToStart(&funcOp.getFunctionBody().front());
+                        rewriter.setInsertionPoint(callOp);
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp:186]: Yes so the problem is when you have a for loop with a call inside (e.g vmap). Here you go outside of the loop and set the insertion point there, then the parameter count is not accessible in the call op in the for loop. So the solution is to set it before the call op.

[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp:186]: Nice catch, that makes sense!
I guess having it at the top-level will be much more efficient (can be reused for multiple calls, doesn't execute repeatedly inside a loop), but maybe we can could achieve that efficiency with a loop-hoisting pass 👍

[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp:186]:
```suggestion
                        # TODO: optimize the placement of the param count call (e.g. loop hoisting)
                        rewriter.setInsertionPoint(callOp);
```

**Comment by dime10** [comment]
Testing this fix on the following example results in very long (infinite?) compile or runtime:

```py
import pennylane as qml
import jax
from jax import numpy as jnp
import catalyst

n_wires = 5
data = jnp.sin(jnp.mgrid[-2:2:0.2].reshape(n_wires, -1)) ** 3
targets = jnp.array([-0.2, 0.4, 0.35, 0.2])

dev_name = "lightning.qubit"
dev = qml.device(dev_name, wires=n_wires)

@qml.qnode(dev)
def circuit(data, weights):
    """Quantum circuit ansatz"""

    for i in range(n_wires):
        qml.RY(data[i], wires=i)

    for i in range(n_wires):
        qml.RX(weights[i, 0], wires=i)
        qml.RY(weights[i, 1], wires=i)
        qml.RX(weights[i, 2], wires=i)
        qml.CNOT(wires=[i, (i + 1) % n_wires])

    return qml.expval(qml.sum(*[qml.PauliZ(i) for i in range(n_wires)]))

# try broadcasting
jit_circuit = catalyst.qjit(catalyst.vmap(circuit, in_axes = (1, None)))

def my_model(data, weights, bias):
    # works with default.qubit
    if dev_name == "default.qubit":
        return circuit(data, weights) + bias

    # works with lightning.qubit, not broadcasted
    # return jnp.array([circuit(jnp.array(d), weights) for d in data.T])

    # only works with loss_fn, fails at grad step
    return jit_circuit(data, weights) + bias

@jax.jit
def loss_fn(params, data, targets):
    predictions = my_model(data, params["weights"], params["bias"])
    loss = jnp.sum((targets - predictions) ** 2 / len(data))
    return loss


weights = jnp.ones([n_wires, 3])
bias = jnp.array(0.)
params = {"weights": weights, "bias": bias}

print(loss_fn(params, data, targets))
print(jax.grad(loss_fn)(params, data, targets))  # runs for > 20 minutes
```

**Comment by rmoyard** [comment]
Closes https://github.com/PennyLaneAI/catalyst/issues/294

**Comment by rmoyard** [comment]
Wow, I cannot believe it will maybe make it to main, after more than two years. Be careful sometimes it causes runtime errors but it also sometimes yields **wrong** results. @dime10 @mehrdad2m  Good luck to the one that will explore the issue!

File: frontend/test/pytest/test_gradient.py [hunk 3]
```diff
@@ -1767,15 +1747,35 @@ def loss_fn(params, data, targets):
...
+        r = weights[..., 1, 2, 0]
+        qml.RY(r, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    weights = jnp.ones([5, 3, 3])
+
+    cat_res = qjit(grad(circuit, argnums=0))(weights)
+    qml.capture.disable()
+    jax_res = jax.grad(circuit, argnums=0)(weights)
+    assert np.allclose(cat_res, jax_res)
```
Comment (JerryChen97) [inline frontend/test/pytest/test_gradient.py]: Does the location of this test also contribute to the original bug?

Comment (mehrdad2m) [reply frontend/test/pytest/test_gradient.py]: haha no. I will fix that. the rebase I think has changed the order, since the PR was very old.
